feat: New options to install packages via micromamba

eitsupi opened the pull request:
Close #7

Implement two options to install packages via micromamba:
- `packages` option to install by the `micromamba install` command.
- `envFile` option to install by the `micromamba create` command.

eitsupi committed:
feat: New option to install packages via micromamba

eitsupi committed:
move install test to the general test

eitsupi committed:
add test

eitsupi commented on test/micromamba/test.sh:
```diff
@@ -7,4 +7,17 @@ source dev-container-features-test-lib
+}
+check "no-conda-forge" no_conda_forge
+
+micromamba install --yes --channel=conda-forge python
```
I don't know why this command fails by non-root users...... (obviously I lack understanding of Conda and Mamba!)
Do we need additional commands to run micromamba non-interactively?

maresb replied:
The user needs to have an initialized shell, i.e. `micromamba activate`. We add this command to `~/.bashrc` [here](https://github.com/mamba-org/devcontainer-features/blob/1a36e997d10589b1a95d732737866c7d504b0ccd/src/micromamba/install.sh#L151). Unfortunately, if `~/.bashrc` doesn't run for whatever reason (shell type not correct, or the user changes), then the environment will not be active.

eitsupi replied:
Thank you for telling me about it.
However, since 7086d5a also failed, it seems that the `bashrc` is loaded correctly and is not the cause of the error.

eitsupi replied:
For non-root users, the packages option also seems to fail by a permission error...
https://github.com/mamba-org/devcontainer-features/actions/runs/3839733691/jobs/6537902861#step:4:2455
This could be a micromamba issue, but I don't know the details.

maresb replied:
This seems strange to me. Unfortunately I don't have any time to look at this now. :(

maresb replied:
The `~/conda-bld` seems to come from the channels config [here](https://github.com/mamba-org/mamba/blob/a94d5ac9d6450c6058d02247fe95f3175e2cab98/libmamba/src/core/channel.cpp#L822). I'm not sure what this means for the problem.

eitsupi replied:
Thank you for taking a look at this.
Fortunately these errors are irrelevant when installing this Feature and using micromamba interactively, so I think we can leave it as is for now.

eitsupi committed:
try source ~/.bashrc

eitsupi committed:
Revert "try source ~/.bashrc"

This reverts commit 7086d5a7a671e1fadbc6a64d2a982a4ce679712c.

eitsupi committed:
fix install command

eitsupi committed:
add a new test

eitsupi committed:
add test for the specFile option

eitsupi committed:
add new options for micromamba create

eitsupi committed:
try to fix permission

to fix the `critical libmamba filesystem error: status: Permission denied [~/conda-bld]` error

eitsupi committed:
Revert "try to fix permission"

This reverts commit 0cf08aa59cfb4159463c11cbe9bf4d38d50f202c.

eitsupi committed:
try to fix permission (take 2)

to fix the critical libmamba filesystem error: status: Permission denied [~/conda-bld] error

eitsupi committed:
Revert "try to fix permission (take 2)"

This reverts commit 17f1ce48da72a496a85c7740062b0b07229146e2.

eitsupi commented:
The errors that are occurring when non-root users try to run `micromamba install` may need to be fixed on the micromamba side.
Changing the directory permission by the `chown` command just before executing these commands did not resolve the errors.

```log
critical libmamba filesystem error: cannot set file time: Operation not permitted [/opt/conda/conda-meta/history]
```

```log
critical libmamba filesystem error: status: Permission denied [~/conda-bld]
```

eitsupi committed:
changing the order of tests and change the base image

eitsupi committed:
fix env file checking

eitsupi commented:
> ```
> critical libmamba filesystem error: cannot set file time: Operation not permitted [/opt/conda/conda-meta/history]
> ```
> 
> ```
> critical libmamba filesystem error: status: Permission denied [~/conda-bld]
> ```

micromamba 1.3 seems to have fixed the first error. The second error continues to occur.

maresb commented:
Hey, thanks for the update! I took a very quick look at this. The activation test seems to be failing due to it running as `root` (where the shell has not been initialized) instead of the `vscode` user. I didn't see anywhere where the `vscode` user was specified. Do we need to add `USER vscode` to the end of thetest-specific Dockerfile?

Regarding `Permission denied [~/conda-bld]` this looks to me like a micromamba bug. For instance it worked after I ran `mkdir ~/conda-bld`, so it's not true that permissions are lacking. I think we can submit a bug report as soon as we come up with a minimal reproducing example.

eitsupi commented:
Thank you for taking a look at this.

> The activation test seems to be failing due to it running as `root` (where the shell has not been initialized) instead of the `vscode` user. I didn't see anywhere where the `vscode` user was specified. Do we need to add `USER vscode` to the end of the test-specific Dockerfile?

Testing with the Dev Container CLI reads `remoteUser` name from the metadata recorded as a label (`devcontainer.metadata`) and executes the script by the remote user, so there should be no need to change the user on the Dockerfile.
I am probably missing something because I am pretty clueless about conda activation.......

> Regarding `Permission denied [~/conda-bld]` this looks to me like a micromamba bug. For instance it worked after I ran `mkdir ~/conda-bld`, so it's not true that permissions are lacking. I think we can submit a bug report as soon as we come up with a minimal reproducing example.

I too think this is a micromamba bug, but I can't explain the situation well and sorry I haven't created an issue yet.
It would be very helpful for you to dig into this.

maresb committed:
Remove test_python from test.sh

We already run test_python in other tests, and installing python is slow

maresb commented:
Hi @eitsupi, I've been trying to understand what's going on here.

I am no longer seeing `Permission denied [~/conda-bld]`, so perhaps it's fixed in v1.3.1?

I am however seeing a "Shell not initialized" error in `test-specfile-option-non-root.sh`. This seems to be due to the test scripts running in a non-interactive Bash shell. Thus the `~/.bashrc` is not being run. One possible ugly workaround is to run

```bash
eval "$(micromamba shell hook --shell=bash)"
```

in the test script if you ever need to run `micromamba activate`.

To explain what's going on, the `micromamba activate` command is supposed to modify the current shell's environment, which it cannot do when executed as a subprocess of the shell. Therefore, the trick is to define a "Bash function" called `micromamba` which overrides the `micromamba` executable. This function passes arguments through to the `micromamba` executable unless the command is `micromamba activate [...]`. (You can look at the output of `micromamba shell hook --shell=bash`, the relevant part being `micromamba() { ... }`.)

eitsupi committed:
try activate micromamba in the test

eitsupi committed:
update test

eitsupi committed:
fix option name in the test

eitsupi committed:
try export root prefix

eitsupi committed:
fix create command

eitsupi committed:
fix create command

eitsupi commented:
@maresb Thanks for making sure the bug is fixed and your advice!
After some more struggle, I think it now works as it's supposed to!

eitsupi committed:
add proposals in all string type options

eitsupi committed:
bump version

eitsupi committed:
add notes about new options

eitsupi committed:
add more test

eitsupi committed:
add link to the user guid

eitsupi committed:
fix test (should set the channels option)

eitsupi committed:
fix test

eitsupi committed:
fix test

maresb committed:
Avoid expanding Bash parameters until the end

maresb committed:
Use login shell with su

maresb committed:
Switch to space separators

maresb committed:
Bump test versions of Micromamba

maresb committed:
Add comment for auto activation

maresb committed:
Resolve the micromamba executable

maresb committed:
Make tests fast

maresb committed:
Explain why we manually run the hook in tests

maresb committed:
Add test-lockfile

maresb commented:
I made several changes and improvements...

* I want the tests to run quickly
* I don't want to test the functionality of Micromamba itself, just the feature. Thus I chose small packages (wget and xz) as opposed to big packages (python, r, julia)
* I made an image which already has the prerequisites so that we don't have to wait for `apt-get update && apt-get install` while running scenarios
* I improved parsing so that we can include specifiers in the packages argument, e.g. `python<3.11`.
* I had to switch to space separators, since comma is a valid character for separators, e.g. `python>=3.10,<3.11`.
* I added a test for lockfiles

Let me know if you disagree with any of these.

Some things I'd like to still do:
* Update the documentation and give some examples of specifiers
* Look into switching from `micromamba create` to `micromamba install` so that we can install from specfiles into the base environment.

eitsupi commented:
Thanks for taking this over!

I think all of these are very good, but I am concerned about the effort it will take to maintain the special base image.
Could we use `mcr.microsoft.com/devcontainers/base:debian` here?
This image has a non-root user, but for scenario testing, we should be able to test the root user as well by specifying like `"remoteUser": "root"`.

eitsupi commented:
> I think I should add a comment by each reference to `ghcr.io/maresb/docker-debian-curl` to clarify that my image can be safely replaced by `debian:stable-slim`.

Since the test is run on CI with the default options, `debian:stable-slim` is not necessary to use it for scenario testing.

https://github.com/mamba-org/devcontainer-features/blob/1a36e997d10589b1a95d732737866c7d504b0ccd/.github/workflows/test.yaml#L36-L41

> [_Oops, I see now that this is a `.json` file, so I probably can't add comments. 😞_]

I believe these are actually jsonc files so perhaps comments can be added.

---

I couldn't be sure because I didn't have time to look into it at the time I made my comment the other day, but I believe that `buildpack-deps:curl` (Docker Official Image) is exactly a minimal image that `bzip2` and `curl` are installed.
How about using this?
https://hub.docker.com/_/buildpack-deps
https://github.com/docker-library/buildpack-deps

maresb commented:
> Since the test is run on CI with the default options, debian:stable-slim is not necessary to use it for scenario testing.

To clarify, I was using the [Dev Container CLI](https://github.com/devcontainers/cli) `devcontainer features test` command to run the tests locally, and they were running slowly. That's why I'm trying to speed up the testing. (Especially since Bash often has such strange behavior!)

> I believe that `buildpack-deps:curl` (Docker Official Image) is exactly a minimal image that bzip2 and curl are installed.

Ah, that looks really promising, but I just checked, and unfortunately bzip2 is missing.

But anyways, I'm really excited that we will now support specifiers and lockfiles. This will be very powerful! :rocket: 

If you manage to find the time, please feel free to finish this up. Regarding the test infrastructure, I think the details are not so important as long as we're doing the right tests. Feel free to revert my stuff if it makes things easier.

eitsupi commented:
> Ah, that looks really promising, but I just checked, and unfortunately bzip2 is missing.

Oh, sorry. It seems bzip2 is only installed in `buildpack-deps:scm`...

I am in no hurry to merge this, but I see no problem with merging it as it is now and making improvements later.
If you want to make updates at this time, feel free to do so at your own pace.

maresb commented:
I agree that it's currently mergeable.

I'll merge and create two issues for the todos in https://github.com/mamba-org/devcontainer-features/pull/11#issuecomment-1465313667. 

Would you like to make a new release?

eitsupi commented:
New version is released! https://github.com/mamba-org/devcontainer-features/releases/tag/feature_micromamba_1.1.0
